cmd/zrepl/mgmt_conn.c

```diff
@@ -0,0 +1,907 @@
+	zvol_io_hdr_t *hdrp = conn->conn_hdr;
+	void *payload = conn->conn_buf;
+	size_t payload_size = conn->conn_bufsiz;
+	zvol_info_t *zinfo;
+	int rc = 0;
+
+	conn->conn_hdr = NULL;
+	conn->conn_buf = NULL;
+	conn->conn_bufsiz = 0;
+	conn->conn_procn = 0;
+
+	switch (hdrp->opcode) {
+	case ZVOL_OPCODE_HANDSHAKE:
+	case ZVOL_OPCODE_PREPARE_FOR_REBUILD:
+	case ZVOL_OPCODE_REPLICA_STATUS:
+		if (payload_size == 0 || payload_size > MAX_NAME_LEN) {
+			rc = reply_error(conn, ZVOL_OP_STATUS_FAILED,
+			    hdrp->opcode, hdrp->io_seq, CS_INIT);
+			break;
+		}
+		strncpy(zvol_name, payload, payload_size);
+		zvol_name[payload_size] = '\0';
+
+		if ((zinfo = uzfs_zinfo_lookup(zvol_name)) == NULL) {
+			fprintf(stderr, "Unknown zvol: %s\n", zvol_name);
+			rc = reply_error(conn, ZVOL_OP_STATUS_FAILED,
+			    hdrp->opcode, hdrp->io_seq, CS_INIT);
+			break;
+		}
+
+		if (hdrp->opcode == ZVOL_OPCODE_HANDSHAKE) {
+			DBGCONN(conn, "Handshake command for zvol %s",
+			    zvol_name);
+			rc = uzfs_zvol_mgmt_do_handshake(conn, hdrp, zvol_name,
+			    zinfo);
+		} else if (hdrp->opcode == ZVOL_OPCODE_PREPARE_FOR_REBUILD) {
+			DBGCONN(conn, "Prepare for rebuild command for zvol %s",
+			    zvol_name);
+			rc = uzfs_zvol_mgmt_do_handshake(conn, hdrp, zvol_name,
+			    zinfo);
+		} else if (hdrp->opcode == ZVOL_OPCODE_REPLICA_STATUS) {
+			DBGCONN(conn, "Replica status command for zvol %s",
+			    zvol_name);
+			rc = uzfs_zvol_rebuild_status(conn, hdrp, zvol_name,
+			    zinfo);
+		} else {
+			ASSERT(0);
+		}
+		uzfs_zinfo_drop_refcnt(zinfo, B_FALSE);
+		break;
+
+
+	case ZVOL_OPCODE_START_REBUILD:
+		/* iSCSI controller will send this msg to downgraded replica */
+		if (payload_size < sizeof (mgmt_ack_t)) {
+			rc = reply_error(conn, ZVOL_OP_STATUS_FAILED,
+			    hdrp->opcode, hdrp->io_seq, CS_INIT);
+			break;
+		}
+		DBGCONN(conn, "Rebuild start command");
+		rc = uzfs_zvol_rebuild_dw_replica_start(conn, hdrp, payload,
+		    payload_size / sizeof (mgmt_ack_t));
+		break;
+
+	default:
+		DBGCONN(conn, "Message with unknown OP code %d", hdrp->opcode);
+		rc = reply_error(conn, ZVOL_OP_STATUS_FAILED, hdrp->opcode,
+		    hdrp->io_seq, CS_INIT);
+		break;
+	}
+	kmem_free(hdrp, sizeof (*hdrp));
+	if (payload != NULL)
+		kmem_free(payload, payload_size);
+
+	return (rc);
+}
+
+/*
+ * Transition to the next state. This is called only if IO buffer was fully
+ * read or written.
+ */
+static int
+move_to_next_state(uzfs_mgmt_conn_t *conn)
+{
+	struct epoll_event ev;
+	zvol_io_hdr_t *hdrp;
+	uint16_t vers;
+	int rc = 0;
+
+	ASSERT3P(conn->conn_bufsiz, ==, conn->conn_procn);
+
+	switch (conn->conn_state) {
+	case CS_CONNECT:
+		DBGCONN(conn, "Connected");
+		/* Fall-through */
+	case CS_INIT:
+		DBGCONN(conn, "Reading version..");
+		if (conn->conn_buf != NULL)
+			kmem_free(conn->conn_buf, conn->conn_bufsiz);
+		conn->conn_buf = kmem_alloc(sizeof (uint16_t), KM_SLEEP);
+		conn->conn_bufsiz = sizeof (uint16_t);
+		conn->conn_procn = 0;
+		ev.events = EPOLLIN;
```

Comment:
EPOLLRDHUP also would be useful

Reply:
```EPOLLRDHUP (since Linux 2.6.17)
              Stream socket peer closed connection, or shut down writing
              half of connection.  (This flag is especially useful for writ‐
              ing simple code to detect peer shutdown when using Edge Trig‐
              gered monitoring.)
```
Firstly our algorithm handles the connection close from the other side reliably so no need to use less known epoll flags like this one. Secondly we don't use edge triggered events which are the reason why this flag was introduced.